### PR TITLE
Introduce ViewModel Injection

### PR DESCRIPTION
**Pull request by d4rken**
## Description
This PR adds the boilerplate necessary for combining ViewModels and dependency injection. As guide/demo the `TestRiskLevelCalculationFragment` has been refactored.

Our injected "ViewModels" are called `VDC` (View Data Controller/Component) because "ViewModel" is a confusing name and overloading the existing `viewModels` extensions would lead to confusion.

Using a `VDC` requires 3 things:
**In a dagger `Module` we need to add a `VDC.Factory` to a lookup map:**
```kotlin
@Module
abstract class TestRiskLevelCalculationFragmentModule {
    @Binds
    @IntoMap
    @VDCKey(TestRiskLevelCalculationFragmentVDC::class)
    abstract fun testRiskLevelFragment(factory: TestRiskLevelCalculationFragmentVDC.Factory): VDCFactory<out VDC>
}
```

**In the `Fragment` we need to instantiate it**
```kotlin
    @Inject lateinit var vdcSource: VDCSource.Factory
    private val vdc: TestRiskLevelCalculationFragmentVDC by vdcsAssisted(
        { vdcSource },
        { factory, handle ->
            factory as TestRiskLevelCalculationFragmentVDC.Factory
            factory.create(handle, navArgs.exampleArgument)
        }
    )
```

* `vdcSource` is a factory that has access to the lookup map we created in step one.
* `vdcsAssisted` is a little indirection that allows us to have an injected class with manual arguments, meaning we can have both injected parameters from our DI graph, and manual parameters, such as fragment/nav graph arguments.

**Finally we need the `VDC` (ViewModel) itself:**
```kotlin
class TestRiskLevelCalculationFragmentVDC @AssistedInject constructor(
    @Assisted private val handle: SavedStateHandle,
    @Assisted private val exampleArg: String?,
    private val context: Context, // App context injected
) : VDC() {

    init {
        Timber.d("VDC: %s", this)
        Timber.d("SavedStateHandle: %s", handle)
        Timber.d("Example arg: %s", exampleArg)
    }

    @AssistedInject.Factory
    interface Factory : VDCFactory<TestRiskLevelCalculationFragmentVDC> {
        fun create(
            handle: SavedStateHandle,
            exampleArg: String?
        ): TestRiskLevelCalculationFragmentVDC
    }
}
```
* Note the little differences to normal Dagger DI, i.e. `AssistedInject`,and `Assisted`. Further details: https://github.com/square/AssistedInject
* `SavedStateHandle` is a framework class that offers behavior similar to `onSavedInstance`.

**Now use the vdc/viewmodel in your fragment by moving logic into it, and keeping UI updates out of it, i.e.**
```kotlin
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        binding.buttonResetRiskLevel.setOnClickListener { vdc.resetRiskLevel() }
        vdc.riskLevelResetEvent.observe2(this) {
            Toast.makeText(
                requireContext(), "Reset done, please fetch diagnosis keys from server again",
                Toast.LENGTH_SHORT
            ).show()
        }
    }
```

## How to test
* Launch the app and use the TestRiskLevelCalculationFragment
* Review the code

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/LiveDataExtensions.kt
Comment (jakobmoellersap) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/LiveDataExtensions.kt:8]: function names with numbers are not great to read imo and we should try to make the naming more explicit

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt
Comment (jakobmoellersap) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt:11]: I think the implementation should be called factory as well, even though it is our source we should still stick to pattern, especially when customizing already generic code.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt
Comment (jakobmoellersap) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt:33]: is there not an option to just use the handle as optional parameter and then use a generic factory?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/SingleLiveEvent.kt
Comment (jakobmoellersap) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/SingleLiveEvent.kt:40]: Shouldn't this still be LiveData? calling it event is imo not correct since an event happens exactly once. However this can be triggered multiple times, e.g. through multiple setValue() calls. Just my opinion, no high prio.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/di/AppInjector.kt
Comment (jakobmoellersap) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/di/AppInjector.kt:25]: should we branch out depending on API version?

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt
Comment (jakobmoellersap) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt:117]: shouldn't we handle a null value more explicit?

**Review by jakobmoellerdev — COMMENTED** [review]
Minor Remarks in the code. Overall I think we should evaluate whether the gain of using pure constructor-based DI is worth the overhead instead of using lateinit vars.

Waiting for more reviews to give approval 👍

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/LiveDataExtensions.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/LiveDataExtensions.kt:8]: This is not necessary to be used, but codesugar that makes code look much nicer. I usually use this naming to denote, that a function is functionally the same, but an upgrade. Autocompletion would show this variant to someone trying to use `.observe`.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt:11]: So `CWAViewModelFactory.Factory` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/SingleLiveEvent.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/ui/SingleLiveEvent.kt:40]: I think `Data` would be missleading here because the data (or rather event here) is consumed.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModelSource.kt:33]: These default factories are used when a viewmodel doesn't need extra assisted arguments, e.g. so the fragment can have

```kotlin
class Fragment .... {
    @AssistedInject.Factory
    interface Factory : SavedStateVDCFactory<GeneratorsFragmentVDC>

}
```

instead of 

```kotlin
class StrawberryViewModel .... {
    @AssistedInject.Factory
    interface Factory : SavedStateVDCFactory<GeneratorsFragmentVDC> {
         fun create(handle: SavedStateHandle): StrawberryViewModel
    }

}
```

Default parameters wouldn't work as it would then still require any user of `SimpleCWAViewModelFactory` to explicitly implement that parameter, i.e.

```kotlin
class ViewModel(private val iDontNeedTheSavedState: SavedStateHandle?)
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/di/AppInjector.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/di/AppInjector.kt:25]: IMHO, no, as we would have different initialization behaviors depending on API version and could have bugs just on some  devices, increasing the chance for bugs that are not easily noticed/reproduced in tests. Too risky too little gain...

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt
Comment (d4rken) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt:117]: It can never be null as we initialize with a value and the data type is not nullable:
```kotlin
val riskScoreState = MutableLiveData<RiskScoreState>(RiskScoreState())
```

**Review by jakobmoellerdev — APPROVED** [review]

**Pull request merged**
